From 89b2f490570462a361d59bf8b3e3b8e539a90af7 Mon Sep 17 00:00:00 2001
From: Nathan Muggli <nmuggli@google.com>
Date: Fri, 12 Jul 2024 13:27:23 -0600
Subject: [PATCH] Revert "[Code Health] Add deprecation comment for
 base::SupportsWeakPtr."

This reverts commit 8780ea75650d3428a18b5526dfaa6ddc46cce446.

Revert "[Code Health] Clean up a friend declaration for base::SupportsWeakPtr."
This reverts commit dd42a71cd065bf2a32ad6e7a58a23b1c48ce132b.

This patch reverts two different commits to weak_ptr.h.
Remove this patch when all uses of SupportsWeakPtr have been removed.
---
 base/memory/weak_ptr.h           | 48 ++------------------------------
 base/memory/weak_ptr_unittest.cc | 11 --------
 2 files changed, 3 insertions(+), 56 deletions(-)

diff --git a/base/memory/weak_ptr.h b/base/memory/weak_ptr.h
index 9c4191bab5..17f9ca6258 100644
--- a/base/memory/weak_ptr.h
+++ b/base/memory/weak_ptr.h
@@ -92,27 +92,6 @@ class ProcessNodeImpl;
 class WorkerNodeImpl;
 }  // namespace performance_manager
 
-// TODO(crbug.com/40485134): This is a hack to prevent new uses of
-// SupportsWeakPtr. The remaining uses are troublesome to fix, so we declare
-// these classes as friends so they have access to the constructor.
-namespace gl {
-class GLContext;
-class GLSurface;
-}  // namespace gl
-
-namespace storage {
-class CopyOrMoveHookDelegate;
-}  // namespace storage
-
-namespace ui {
-class MenuModel;
-class TextInputClient;
-}  // namespace ui
-
-namespace views {
-class WidgetDelegate;
-}  // namespace views
-
 namespace base {
 
 namespace sequence_manager::internal {
@@ -487,23 +466,16 @@ class WeakPtrFactory : public internal::WeakPtrFactoryBase {
   }
 };
 
-// TODO(crbug.com/40485134): This is a hack to prevent new uses of
-// SupportsWeakPtr. The remaining uses are troublesome to fix, so we declare
-// these classes as friends so they have access to the constructor. This
-// class is used in the unit tests.
-namespace weak_ptr_unittest {
-struct Target;
-}
-
 // A class may extend from SupportsWeakPtr to let others take weak pointers to
 // it. This avoids the class itself implementing boilerplate to dispense weak
 // pointers.  However, since SupportsWeakPtr's destructor won't invalidate
-// weak pointers to the class until after the derived class's members have been
+// weak pointers to the class until after the derived class' members have been
 // destroyed, its use can lead to subtle use-after-destroy issues.
-// This class cannot be used in new code. See crbug.com/40485134.
 template <class T>
 class SupportsWeakPtr : public internal::SupportsWeakPtrBase {
  public:
+  SupportsWeakPtr() = default;
+
   SupportsWeakPtr(const SupportsWeakPtr&) = delete;
   SupportsWeakPtr& operator=(const SupportsWeakPtr&) = delete;
 
@@ -515,20 +487,6 @@ class SupportsWeakPtr : public internal::SupportsWeakPtrBase {
   ~SupportsWeakPtr() = default;
 
  private:
-  friend struct MultiplyDerivedProducer;
-  friend struct Producer;
-  friend struct weak_ptr_unittest::Target;
-  friend class gl::GLContext;
-  friend class gl::GLSurface;
-  friend class storage::CopyOrMoveHookDelegate;
-  friend class ui::MenuModel;
-  friend class ui::TextInputClient;
-  friend class views::WidgetDelegate;
-
-  // The constructor is private so only the declared friends can construct
-  // instances.
-  SupportsWeakPtr() = default;
-
   internal::WeakReferenceOwner weak_reference_owner_;
 };
 
diff --git a/base/memory/weak_ptr_unittest.cc b/base/memory/weak_ptr_unittest.cc
index bd5d7c4d82..02a44772c1 100644
--- a/base/memory/weak_ptr_unittest.cc
+++ b/base/memory/weak_ptr_unittest.cc
@@ -64,20 +64,9 @@ struct Base {
 struct Derived : public Base {};
 
 struct TargetBase {};
-
-}  // namespace
-
-// TODO(crbug.com/40485134): Target needs to be visible to declare
-// it as a friend of SupportsWeakPtr.
-namespace weak_ptr_unittest {
 struct Target : public TargetBase, public SupportsWeakPtr<Target> {
   virtual ~Target() = default;
 };
-}  // namespace weak_ptr_unittest
-
-namespace {
-
-using weak_ptr_unittest::Target;
 
 struct DerivedTarget : public Target {};
 
-- 
2.45.2.1089.g2a221341d9-goog

